In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import re
import requests
from bs4 import BeautifulSoup
from selenium import webdriver
import time

In [5]:
URL = 'https://www.flipkart.com/search?q=phone&otracker=search&otracker1=search&marketplace=FLIPKART&as-show=on&as=off&p%5B%5D=facets.price_range.from%3D20000&p%5B%5D=facets.price_range.to%3DMax&page=1'
page = requests.get(URL)
page.status_code

403

### Without Headers
Your request looks like:

“Hi, I am a script. Definitely not a bot. Trust me.”

Flipkart:

“403. Go away.”

### What is User-Agent?
When you visit a website, your browser sends info like:

Browser name (Chrome, Firefox)

OS (Windows, Mac)
Device type

Example:
Chrome on Windows → Mozilla/5.0 ..

### With headers:

Now your request says:

“Hello, I am Chrome browser on Windows.”

Flipkart:

“Alright, come in.”

In [3]:

URL = 'https://www.flipkart.com/search?q=phone&otracker=search&otracker1=search&marketplace=FLIPKART&as-show=on&as=off&p%5B%5D=facets.price_range.from%3D20000&p%5B%5D=facets.price_range.to%3DMax&page=1'

headers = {
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/115.0 Safari/537.36"
}

page = requests.get(URL, headers=headers)

print(page.status_code)

200


In [4]:
#It returns the raw data (in bytes) that the server sent back.
# page.content

In [5]:
soup = BeautifulSoup(page.content, 'html.parser')
# print(soup.prettify())

In [6]:
name = soup.find('div', class_ = 'RG5Slk')
name

<div class="RG5Slk">IQOO Z11x 5G (Prismatic Green, 128 GB)</div>

In [7]:
name.text

'IQOO Z11x 5G (Prismatic Green, 128 GB)'

In [8]:
price = soup.find('div', class_ = 'hZ3P6w DeU9vF')
price.text

'₹21,659'

In [9]:
rating = soup.find('div', class_ = 'MKiFS6')
rating.text

'4.4'

In [10]:
ratings_reviews = soup.find('span', class_ = 'PvbNMB')
ratings_reviews.text

'464 Ratings\xa0&\xa040 Reviews'

In [11]:
specifications = soup.find('div', class_ = 'CMXw7N')
specifications

<div class="CMXw7N"><ul class="HwRTzP"><li class="DTBslk">6 GB RAM | 128 GB ROM</li><li class="DTBslk">17.02 cm (6.7 inch) Display</li><li class="DTBslk">50MP Rear Camera</li><li class="DTBslk">7200 mAh Battery</li><li class="DTBslk">12 Months Brand Warranty</li></ul></div>

In [12]:
specs = specifications.find_all('li', class_ = 'DTBslk')
print(specs)
a=[]
for i in specs:
    a.append(i.text)
a

[<li class="DTBslk">6 GB RAM | 128 GB ROM</li>, <li class="DTBslk">17.02 cm (6.7 inch) Display</li>, <li class="DTBslk">50MP Rear Camera</li>, <li class="DTBslk">7200 mAh Battery</li>, <li class="DTBslk">12 Months Brand Warranty</li>]


['6 GB RAM | 128 GB ROM',
 '17.02 cm (6.7 inch) Display',
 '50MP Rear Camera',
 '7200 mAh Battery',
 '12 Months Brand Warranty']

In [13]:
print(len(specs))
print(specs[0].text)
print(specs[1].text)
print(specs[2].text)
print(specs[3].text)
print(specs[4].text)
print(specs[5].text)

5
6 GB RAM | 128 GB ROM
17.02 cm (6.7 inch) Display
50MP Rear Camera
7200 mAh Battery
12 Months Brand Warranty


IndexError: list index out of range

In [ ]:
product_details = soup.find_all("div",class_="ZFwe0M row")
print(len(product_details))
# product_details

In [ ]:
for product in product_details:
    #name :
    name = product.find('div', class_ = 'RG5Slk')
    if name is None:
        print("Name is not available")
    else:
        print(name.text)
        #price
    price = product.find('div', class_ = 'hZ3P6w DeU9vF')
    if price is None:
        print("price is not available")
    else:
        print(price.text)
    # ratings
    rating = product.find('div', class_ = 'MKiFS6')
    if rating is None:
        print("rating is not available")
    else:
        print(rating.text)
    #ratings and reviews
    ratings_reviews = soup.find('span', class_ = 'PvbNMB')
    if ratings_reviews is None:
        print("rating_reviews is not available")
    else:
        print(ratings_reviews.text)
    #specs
    features = []
    specifications = product.find('div', class_ = 'CMXw7N')
    specs = specifications.find_all('li', class_ = 'DTBslk')
    for feature in specs:
        features.append(feature.text)
    print(features)
    print("="*100)
    print()

In [ ]:
titles =[]
for product in product_details:
    #name :
    name = product.find('div', class_ = 'RG5Slk')
    if name is None:
        titles.append(np.nan)
    else:
    
        titles.append(name.text)
titles

In [15]:
# Storing information in lists
titles = []
prices = []
ratings = []
RnR = []
features = []
pgno = []

# loop to go through webpages
for x in range(1,11):
    start_time = time.time()
    
    URL = 'https://www.flipkart.com/search?q=phone&otracker=search&otracker1=search&marketplace=FLIPKART&as-show=on&as=off&p%5B%5D=facets.price_range.from%3D20000&p%5B%5D=facets.price_range.to%3DMax&page=1'.format(x)
    
    headers = {
        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/115.0 Safari/537.36"
    }
    
    page = requests.get(URL, headers=headers)
    soup = BeautifulSoup(page.content, 'html.parser')
    
    # Extracting product details of all products on the webpage
    product_details = soup.find_all('div', class_ = 'ZFwe0M row')

    for product in product_details:
        # name
        name = product.find('div', class_ = 'RG5Slk')

        if name is None:
            titles.append(np.nan)
        else:
            titles.append(name.text)

        # price
        price = product.find('div', class_ = 'hZ3P6w DeU9vF')
        if price is None:
            prices.append(np.nan)
        else:
            prices.append(price.text)

        # rating
        rating = product.find('div', class_ = 'MKiFS6')
        if rating is None:
            ratings.append(np.nan)
        else:
            ratings.append(rating.text)

        # ratings and reviews
        ratings_reviews = product.find('span', class_ = 'PvbNMB')
        if ratings_reviews is None:
            RnR.append(np.nan)
        else:
            RnR.append(ratings_reviews.text)

        # specifications
        f = []
        specifications = product.find('div', class_ = 'CMXw7N')
        specs = specifications.find_all('li', class_ = 'DTBslk')
        for feature in specs:
            f.append(feature.text)
        features.append(f)
        
        # page number
        pgno.append(x)
        
    end_time = time.time()
    print('Page {} completed in {} seconds'.format(x, end_time - start_time))

Page 1 completed in 0.8784656524658203 seconds
Page 2 completed in 1.497955322265625 seconds
Page 3 completed in 1.076066017150879 seconds
Page 4 completed in 0.9753823280334473 seconds
Page 5 completed in 1.3976051807403564 seconds
Page 6 completed in 0.9952309131622314 seconds
Page 7 completed in 1.1727080345153809 seconds
Page 8 completed in 3.916471481323242 seconds
Page 9 completed in 1.2857615947723389 seconds
Page 10 completed in 0.9276559352874756 seconds


In [16]:
data = {'Product_name' : titles,
        'Price'        : prices,
        'Rating'       : ratings,
        'No_of_RnR'    : RnR,
        'Features'     : features,
        'PageNo'       : pgno}

df = pd.DataFrame(data)
df.head(10)

,Product_name,Price,Rating,No_of_RnR,Features,PageNo
0,"IQOO Z11x 5G (Prismatic Green, 128 GB)","₹21,659",4.4,464 Ratings & 40 Reviews,"[6 GB RAM | 128 GB ROM, 17.02 cm (6.7 inch) Di...",1
1,MOTOROLA Edge 60 Fusion 5G (PANTONE Slipstream...,"₹22,999",4.4,"1,19,085 Ratings & 7,005 Reviews","[8 GB RAM | 128 GB ROM | Expandable Upto 1 TB,...",1
2,"IQOO 15R 5G (Dark Knight, 256 GB)","₹45,988",4.7,100 Ratings & 15 Reviews,"[8 GB RAM | 256 GB ROM, 16.74 cm (6.59 inch) D...",1
3,"IQOO Z11x 5G (Titan Black, 128 GB)","₹23,368",4.4,692 Ratings & 61 Reviews,"[8 GB RAM | 128 GB ROM, 17.02 cm (6.7 inch) Di...",1
4,"OnePlus Nord CE6 Lite (Hyper Black, 128 GB)","₹23,925",3.4,10 Ratings & 2 Reviews,[6 GB RAM | 128 GB ROM | Expandable Upto 128 G...,1
5,REDMI Note 15 SE 5G Snapdragon (Crimson Reserv...,"₹22,999",4.3,"1,456 Ratings & 166 Reviews","[6 GB RAM | 128 GB ROM | Expandable Upto 1 TB,...",1
6,"IQOO Z11x 5G (Prismatic Green, 128 GB)","₹22,988",4.4,692 Ratings & 61 Reviews,"[8 GB RAM | 128 GB ROM, 17.02 cm (6.7 inch) Di...",1
7,"REDMI Note 15 SE 5G (Crimson Reserve, 256 GB)","₹26,999",4.3,722 Ratings & 85 Reviews,"[8 GB RAM | 256 GB ROM | Expandable Upto 1 TB,...",1
8,"IQOO 15R 5G (Triumph Silver, 256 GB)","₹45,988",4.7,100 Ratings & 15 Reviews,"[8 GB RAM | 256 GB ROM, 16.74 cm (6.59 inch) D...",1
9,"realme P4x 5G (Matte Silver, 256 GB)","₹20,999",4.3,"5,033 Ratings & 383 Reviews","[8 GB RAM | 256 GB ROM, 17.07 cm (6.72 inch) F...",1


In [17]:
# Converting price and rating to float
df.Rating = df.Rating.astype(float)
# df.Price = df.Price.apply(lambda s : float(s.replace('₹','').replace(',',''))
df.Price = df.Price.str.replace('₹','').str.replace(',','').astype(float)

print(df.Rating.dtype)
print(df.Price.dtype)
df.head()

float64
float64


,Product_name,Price,Rating,No_of_RnR,Features,PageNo
0,"IQOO Z11x 5G (Prismatic Green, 128 GB)",21659.0,4.4,464 Ratings & 40 Reviews,"[6 GB RAM | 128 GB ROM, 17.02 cm (6.7 inch) Di...",1
1,MOTOROLA Edge 60 Fusion 5G (PANTONE Slipstream...,22999.0,4.4,"1,19,085 Ratings & 7,005 Reviews","[8 GB RAM | 128 GB ROM | Expandable Upto 1 TB,...",1
2,"IQOO 15R 5G (Dark Knight, 256 GB)",45988.0,4.7,100 Ratings & 15 Reviews,"[8 GB RAM | 256 GB ROM, 16.74 cm (6.59 inch) D...",1
3,"IQOO Z11x 5G (Titan Black, 128 GB)",23368.0,4.4,692 Ratings & 61 Reviews,"[8 GB RAM | 128 GB ROM, 17.02 cm (6.7 inch) Di...",1
4,"OnePlus Nord CE6 Lite (Hyper Black, 128 GB)",23925.0,3.4,10 Ratings & 2 Reviews,[6 GB RAM | 128 GB ROM | Expandable Upto 128 G...,1


In [18]:
text = '3,346 Ratings & 223 Reviews'
rating_regex = re.compile(r'([\d,\s]+)Ratings')
ratings = rating_regex.findall(text)
int(ratings[0].strip().replace(',',''))

3346

In [19]:
# creating functions to get number of ratings and number of reviews
def get_ratings(text):
    if text is np.nan:
        return 0
    else:
        rating_regex = re.compile(r'([\d,\s]+)Ratings')
        ratings = rating_regex.findall(text)
        ratings = int(ratings[0].strip().replace(',',''))
        return ratings
    
def get_reviews(text):
    if text is np.nan:
        return 0
    else:
        review_regex = re.compile(r'([\d,\s]+)Reviews')
        reviews = review_regex.findall(text)
        reviews = int(reviews[0].strip().replace(',',''))
        return reviews
    
print(get_ratings('3,346 Ratings & 223 Reviews'))
print(get_reviews('3,346 Ratings & 223 Reviews'))

3346
223


In [20]:
# Creating columns No_of_ratings and No_of_reviews
df['No_of_ratings'] = df.No_of_RnR.apply(get_ratings)
df['No_of_reviews'] = df.No_of_RnR.apply(get_reviews)
df.drop('No_of_RnR', axis = 1, inplace = True)

df.head(10)

,Product_name,Price,Rating,Features,PageNo,No_of_ratings,No_of_reviews
0,"IQOO Z11x 5G (Prismatic Green, 128 GB)",21659.0,4.4,"[6 GB RAM | 128 GB ROM, 17.02 cm (6.7 inch) Di...",1,464,40
1,MOTOROLA Edge 60 Fusion 5G (PANTONE Slipstream...,22999.0,4.4,"[8 GB RAM | 128 GB ROM | Expandable Upto 1 TB,...",1,119085,7005
2,"IQOO 15R 5G (Dark Knight, 256 GB)",45988.0,4.7,"[8 GB RAM | 256 GB ROM, 16.74 cm (6.59 inch) D...",1,100,15
3,"IQOO Z11x 5G (Titan Black, 128 GB)",23368.0,4.4,"[8 GB RAM | 128 GB ROM, 17.02 cm (6.7 inch) Di...",1,692,61
4,"OnePlus Nord CE6 Lite (Hyper Black, 128 GB)",23925.0,3.4,[6 GB RAM | 128 GB ROM | Expandable Upto 128 G...,1,10,2
5,REDMI Note 15 SE 5G Snapdragon (Crimson Reserv...,22999.0,4.3,"[6 GB RAM | 128 GB ROM | Expandable Upto 1 TB,...",1,1456,166
6,"IQOO Z11x 5G (Prismatic Green, 128 GB)",22988.0,4.4,"[8 GB RAM | 128 GB ROM, 17.02 cm (6.7 inch) Di...",1,692,61
7,"REDMI Note 15 SE 5G (Crimson Reserve, 256 GB)",26999.0,4.3,"[8 GB RAM | 256 GB ROM | Expandable Upto 1 TB,...",1,722,85
8,"IQOO 15R 5G (Triumph Silver, 256 GB)",45988.0,4.7,"[8 GB RAM | 256 GB ROM, 16.74 cm (6.59 inch) D...",1,100,15
9,"realme P4x 5G (Matte Silver, 256 GB)",20999.0,4.3,"[8 GB RAM | 256 GB ROM, 17.07 cm (6.72 inch) F...",1,5033,383


In [21]:
text = 'CMF by Nothing Phone 2 Pro'
brand_regex = re.compile(r"^\w+") # find the first word from the beginning of string...
brand_regex.findall(text)[0]

'CMF'

In [22]:
df["Brand"] = df.Product_name.apply(lambda s : brand_regex.findall(s)[0])
df.head(5)

,Product_name,Price,Rating,Features,PageNo,No_of_ratings,No_of_reviews,Brand
0,"IQOO Z11x 5G (Prismatic Green, 128 GB)",21659.0,4.4,"[6 GB RAM | 128 GB ROM, 17.02 cm (6.7 inch) Di...",1,464,40,IQOO
1,MOTOROLA Edge 60 Fusion 5G (PANTONE Slipstream...,22999.0,4.4,"[8 GB RAM | 128 GB ROM | Expandable Upto 1 TB,...",1,119085,7005,MOTOROLA
2,"IQOO 15R 5G (Dark Knight, 256 GB)",45988.0,4.7,"[8 GB RAM | 256 GB ROM, 16.74 cm (6.59 inch) D...",1,100,15,IQOO
3,"IQOO Z11x 5G (Titan Black, 128 GB)",23368.0,4.4,"[8 GB RAM | 128 GB ROM, 17.02 cm (6.7 inch) Di...",1,692,61,IQOO
4,"OnePlus Nord CE6 Lite (Hyper Black, 128 GB)",23925.0,3.4,[6 GB RAM | 128 GB ROM | Expandable Upto 128 G...,1,10,2,OnePlus


In [23]:
df.Features = df.Features.apply(lambda x : ",".join(x))
df.head(3)

,Product_name,Price,Rating,Features,PageNo,No_of_ratings,No_of_reviews,Brand
0,"IQOO Z11x 5G (Prismatic Green, 128 GB)",21659.0,4.4,"6 GB RAM | 128 GB ROM,17.02 cm (6.7 inch) Disp...",1,464,40,IQOO
1,MOTOROLA Edge 60 Fusion 5G (PANTONE Slipstream...,22999.0,4.4,"8 GB RAM | 128 GB ROM | Expandable Upto 1 TB,1...",1,119085,7005,MOTOROLA
2,"IQOO 15R 5G (Dark Knight, 256 GB)",45988.0,4.7,"8 GB RAM | 256 GB ROM,16.74 cm (6.59 inch) Dis...",1,100,15,IQOO


In [24]:
df.Features[0]

'6 GB RAM | 128 GB ROM,17.02 cm (6.7 inch) Display,50MP Rear Camera,7200 mAh Battery,12 Months Brand Warranty'

In [25]:
text = df.Features[1]
processor_regex = re.compile(r"([A-Za-z0-9\s]+)Processor")
processor_regex.findall(text)


['Dimensity 7400 ']

In [26]:
def get_processor(text):
    processor_regex = re.compile(r"([A-Za-z0-9\s]+)Processor")
    processor = processor_regex.findall(text)
    if len(processor) == 0:
        return np.nan
    else:
        return processor[0].strip()
get_processor(text)
    
    

'Dimensity 7400'

In [27]:
df["Processor"] = df.Features.apply(get_processor)
df.head(5)

,Product_name,Price,Rating,Features,PageNo,No_of_ratings,No_of_reviews,Brand,Processor
0,"IQOO Z11x 5G (Prismatic Green, 128 GB)",21659.0,4.4,"6 GB RAM | 128 GB ROM,17.02 cm (6.7 inch) Disp...",1,464,40,IQOO,NaN
1,MOTOROLA Edge 60 Fusion 5G (PANTONE Slipstream...,22999.0,4.4,"8 GB RAM | 128 GB ROM | Expandable Upto 1 TB,1...",1,119085,7005,MOTOROLA,Dimensity 7400
2,"IQOO 15R 5G (Dark Knight, 256 GB)",45988.0,4.7,"8 GB RAM | 256 GB ROM,16.74 cm (6.59 inch) Dis...",1,100,15,IQOO,NaN
3,"IQOO Z11x 5G (Titan Black, 128 GB)",23368.0,4.4,"8 GB RAM | 128 GB ROM,17.02 cm (6.7 inch) Disp...",1,692,61,IQOO,NaN
4,"OnePlus Nord CE6 Lite (Hyper Black, 128 GB)",23925.0,3.4,6 GB RAM | 128 GB ROM | Expandable Upto 128 GB...,1,10,2,OnePlus,Qualcomm Snapdragon 7s Gen 4


In [28]:
# CREATING A REGULAR EXPRESSIONTO EXTRACT RAM
text = df.Features[0]
RAM_regex = re.compile(r"([A-Za-z0-9\s]+)RAM")
RAM_regex.findall(text)[0].strip()

'6 GB'

In [29]:
def get_ram(text):
    regex = re.compile(r"([A-Za-z0-9\s]+)RAM")
    result = regex.findall(text)
    if len(result)==0:
        return np.nan
    else:
        return result[0].strip()
get_ram(df.Features[0])

'6 GB'

In [30]:
df["RAM"] = df.Features.apply(get_ram)
df.head(2)

,Product_name,Price,Rating,Features,PageNo,No_of_ratings,No_of_reviews,Brand,Processor,RAM
0,"IQOO Z11x 5G (Prismatic Green, 128 GB)",21659.0,4.4,"6 GB RAM | 128 GB ROM,17.02 cm (6.7 inch) Disp...",1,464,40,IQOO,NaN,6 GB
1,MOTOROLA Edge 60 Fusion 5G (PANTONE Slipstream...,22999.0,4.4,"8 GB RAM | 128 GB ROM | Expandable Upto 1 TB,1...",1,119085,7005,MOTOROLA,Dimensity 7400,8 GB


In [31]:
# CREATING A REGULAR EXPRESSIONTO EXTRACT RAM
text = df.Features[0]
ROM_regex = re.compile(r"([A-Za-z0-9\s]+)ROM")
ROM_regex.findall(text)[0].strip()

'128 GB'

In [32]:
def get_rom(text):
    regex = re.compile(r"([A-Za-z0-9\s]+)ROM")
    result = regex.findall(text)
    if len(result)==0:
        return np.nan
    else:
        return result[0].strip()
get_rom(df.Features[0])

'128 GB'

In [33]:
df["ROM"] = df.Features.apply(get_rom)
df.head(2)

,Product_name,Price,Rating,Features,PageNo,No_of_ratings,No_of_reviews,Brand,Processor,RAM,ROM
0,"IQOO Z11x 5G (Prismatic Green, 128 GB)",21659.0,4.4,"6 GB RAM | 128 GB ROM,17.02 cm (6.7 inch) Disp...",1,464,40,IQOO,NaN,6 GB,128 GB
1,MOTOROLA Edge 60 Fusion 5G (PANTONE Slipstream...,22999.0,4.4,"8 GB RAM | 128 GB ROM | Expandable Upto 1 TB,1...",1,119085,7005,MOTOROLA,Dimensity 7400,8 GB,128 GB


In [34]:
df.RAM.value_counts()

RAM
8 GB     170
6 GB      30
12 GB     30
Name: count, dtype: int64

In [43]:
text = df.Features[0]
battery_regex = re.compile(r"[A-Za-z0-9\s]+mAh")
battery_regex.findall(text)


['7200 mAh']

In [45]:
def get_battery(text):
    regex = re.compile(r"[A-Za-z0-9\s]+mAh")
    result = regex.findall(text)
    if len(result) == 0:
        return np.nan
    else:
        return result[0].strip()
get_battery(text)

'7200 mAh'

In [47]:
df["Battery"] = df.Features.apply(get_battery)
df.head(2)

,Product_name,Price,Rating,Features,PageNo,No_of_ratings,No_of_reviews,Brand,Processor,RAM,ROM,Battery
0,"IQOO Z11x 5G (Prismatic Green, 128 GB)",21659.0,4.4,"6 GB RAM | 128 GB ROM,17.02 cm (6.7 inch) Disp...",1,464,40,IQOO,NaN,6 GB,128 GB,7200 mAh
1,MOTOROLA Edge 60 Fusion 5G (PANTONE Slipstream...,22999.0,4.4,"8 GB RAM | 128 GB ROM | Expandable Upto 1 TB,1...",1,119085,7005,MOTOROLA,Dimensity 7400,8 GB,128 GB,5500 mAh


In [52]:
## DISPLAY....
text = df.Features[0]
display_regex = re.compile(r"[0-9\s.]+inch")
display_regex.findall(text)[0]

'6.7 inch'

In [53]:
def get_display(text):
    regex = re.compile(r"[0-9\s.]+inch")
    result = regex.findall(text)
    if len(result) == 0:
        return np.nan
    else:
        return result[0].strip()
get_display(text)

'6.7 inch'

In [54]:
df["Display_Size"] = df.Features.apply(get_display)
df.head(3)

,Product_name,Price,Rating,Features,PageNo,No_of_ratings,No_of_reviews,Brand,Processor,RAM,ROM,Battery,Display_Size
0,"IQOO Z11x 5G (Prismatic Green, 128 GB)",21659.0,4.4,"6 GB RAM | 128 GB ROM,17.02 cm (6.7 inch) Disp...",1,464,40,IQOO,NaN,6 GB,128 GB,7200 mAh,6.7 inch
1,MOTOROLA Edge 60 Fusion 5G (PANTONE Slipstream...,22999.0,4.4,"8 GB RAM | 128 GB ROM | Expandable Upto 1 TB,1...",1,119085,7005,MOTOROLA,Dimensity 7400,8 GB,128 GB,5500 mAh,6.67 inch
2,"IQOO 15R 5G (Dark Knight, 256 GB)",45988.0,4.7,"8 GB RAM | 256 GB ROM,16.74 cm (6.59 inch) Dis...",1,100,15,IQOO,NaN,8 GB,256 GB,7600 mAh,6.59 inch


In [66]:
text = df.Features[1]
camera_regex = re.compile(r"[A-Za-z0-9\s(\)+|]+Camera")
camera_regex.findall(text)[0]

'50MP + 13MP | 32MP Front Camera'

In [68]:
def get_camera(text):
    regex = re.compile(r"[A-Za-z0-9\s(\)+|]+Camera")
    result = regex.findall(text)
    if len(result)==0:
        return np.nan
    else:
        return result[0].strip()
get_camera(text)


'50MP + 13MP | 32MP Front Camera'

In [69]:
df["Camera"] = df.Features.apply(get_camera)
df.head(2)

,Product_name,Price,Rating,Features,PageNo,No_of_ratings,No_of_reviews,Brand,Processor,RAM,ROM,Battery,Display_Size,Camera
0,"IQOO Z11x 5G (Prismatic Green, 128 GB)",21659.0,4.4,"6 GB RAM | 128 GB ROM,17.02 cm (6.7 inch) Disp...",1,464,40,IQOO,NaN,6 GB,128 GB,7200 mAh,6.7 inch,50MP Rear Camera
1,MOTOROLA Edge 60 Fusion 5G (PANTONE Slipstream...,22999.0,4.4,"8 GB RAM | 128 GB ROM | Expandable Upto 1 TB,1...",1,119085,7005,MOTOROLA,Dimensity 7400,8 GB,128 GB,5500 mAh,6.67 inch,50MP + 13MP | 32MP Front Camera


In [70]:
df.drop(["Features","PageNo"],axis = 1,inplace= True)
df.head(1)

,Product_name,Price,Rating,No_of_ratings,No_of_reviews,Brand,Processor,RAM,ROM,Battery,Display_Size,Camera
0,"IQOO Z11x 5G (Prismatic Green, 128 GB)",21659.0,4.4,464,40,IQOO,NaN,6 GB,128 GB,7200 mAh,6.7 inch,50MP Rear Camera


In [71]:
df.to_csv("Flipkart_phones.csv",index = False)

## `PERFORM EXPLORATORY DATA ANALYSIS